BWIndex.py

```diff
@@ -0,0 +1,239 @@
+			D.node(str(i), label)
+		for tail in self.G:
+			for head in self.G[tail]:
+				D.edge(str(tail), str(head))
+		D.render(outFile, view=True)
+
+	def report(self):
+		print(self.G, self.H, self.NS, self.NE, self.SS, self.BWT, self.suffixArray, self.FirstOccurrence, self.Count)
+
+	def appendNode(self, parent, nodeStringStart, suffixStart):
+		newNode = self.nodeCount
+		self.G[parent] = self.G.get(parent, []) + [newNode]
+		self.H[newNode] = parent
+		self.NS[newNode] = nodeStringStart
+		self.NE[newNode] = len(self.Text) - 1
+		self.SS[newNode] = suffixStart
+		self.nodeCount += 1
+
+	def splitNode(self, node, index):
+		# introduce a new node whose parent is the original parent of node, make it the new parent of node
+		newNode = self.nodeCount
+		parent = self.H[node]
+		self.G[newNode] = [node]
+		self.G[parent].remove(node)
+		self.G[parent] = self.G.get(parent, []) + [newNode]
+		self.H[newNode] = parent
+		self.H[node] = newNode
+		self.NS[newNode] = self.NS[node]
+		self.NE[newNode] = index - 1
+		self.NS[node] = index
+		self.nodeCount += 1
+		return newNode
+
+	def findMatchingChild(self, node, loc):
+		for child in self.G[node]:
+			if self.Text[loc] == self.Text[self.NS[child]]:
+				return child
+		return -1
+
+	def firstMismatch(self, node, ix):
+		offset = 1
+		while True:
+			if ix + offset >= len(self.Text):
+				return self.NS[node] + offset
+			if self.NS[node] + offset > self.NE[node]:
+				return -1
+			if self.Text[self.NS[node] + offset] != self.Text[ix + offset]:
+				return self.NS[node] + offset
+			offset += 1
+
+	def threadSuffix(self, ix):
+		'''
+		look for a node that matches the start of the current substring
+		if none found, create a new node and set its string to the current substring and SS to the initial index
+		if one is found, look for a mismatch within the string
+		if a mismatch is found, split the target node and append the remainder of the substring
+		if no mismatch, advance curIx by the length of the node string and iterate
+		'''
+		node = 0
+		curIx = ix
+		while True:
+			nextNode = self.findMatchingChild(node, curIx)
+			if nextNode == -1:
+				self.appendNode(node, curIx, ix)
+				break
+			node = nextNode
+			splitLoc = self.firstMismatch(node, curIx)
+			if splitLoc == -1:
+				curIx += self.NE[node] - self.NS[node] + 1
+				continue
+			newNode = self.splitNode(node, splitLoc)
+			self.appendNode(newNode, splitLoc + (curIx - self.NS[newNode]), ix)
+			c = self.nodeCount - 1
+			break
+
+	def computeCountArray(self):
+		for i in self.alphabet:
+			self.Count[i] = [0] * (len(self.Text) + 1)  # check does this cause any reference issues?
+		for j in range(1, len(self.BWT) + 1):
+			self.Count[self.BWT[j - 1]][j] = self.Count[self.BWT[j - 1]][j - 1] + 1
+			for k in self.alphabet:
+				if k != self.BWT[j - 1]:
+					self.Count[k][j] = self.Count[k][j - 1]
+		return (self.Count)
+
+	def FindCount(self, Pattern):
+		top = 0
+		bottom = len(self.BWT) - 1
+		while top <= bottom:

```

Comment:
It looks like `None` is returned when `top > bottom`? I assume you want to return something else in that case. Might be worth adding the `len(pattern) > 0` check to the while condition and returning `bottom - top + 1` at the end of the function?